---
title: "Virus detection in Cassava SRA `r params$SRA`"
author: "Marco Chiapello"
date: "28/04/2021"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  SRA: PRJNA297492
  sratoolkit: "/Users/raffaella/Desktop/bioutils/sratoolkit.2.11.0-mac64/bin/fasterq-dump"
  star: "/Users/raffaella/Desktop/bioutils/STAR-2.7.9a/bin/MacOSX_x86_64/STAR"
  bbtools: "/ME4/massimo/software/bbmap/"
  trinity: ""
---

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(fs)
library(tidyverse)
library(DT)
library(biomartr)
library(here)
setwd(here())

# params <- NULL
# params$SRA <- "PRJNA297492"
# params$sratoolkit <-  "/home/massimo/sratoolkit.2.9.6-1-ubuntu64/bin/fasterq-dump"
# params$bbtools <-  "/ME4/massimo/software/bbmap/"
# params$star <- "/ME4/massimo/software/STAR-2.7.8a/bin/Linux_x86_64/STAR"
# params$trinity <- "/usr/share/trinity291/Trinity"
```

# Download needed data

## Download Cassava genome

_Manihot esculenta_ genome has been downloaded from NCBI RefSeq dataset. 

Here all the information about the downloaded genome: 

```{r warning=FALSE, message=FALSE}
if(file_exists("_ncbi_downloads/genomes/Manihot_esculenta_genomic_refseq.fna")){
  suME <- read_tsv("_ncbi_downloads/genomes/doc_Manihot_esculenta_db_refseq.txt") %>%
    separate(`File Name: Manihot_esculenta_genomic_refseq.fna.gz`, into = c("Data", "Info"), sep = ": ")
} else {
  ME.genome.refseq <- getGenome(db       = "refseq", 
                                organism = "Manihot esculenta")
  suME <- read_tsv("_ncbi_downloads/genomes/doc_Manihot_esculenta_db_refseq.txt") %>%
    separate(`File Name: Manihot_esculenta_genomic_refseq.fna.gz`, into = c("Data", "Info"), sep = ": ")
}

datatable(suME)

```


Download from NCBI SRA the run of interest.

### Table of SRA matadata

```{r warning=FALSE, message=FALSE}
# Build the link to download metadata file
system(paste(
  "wget 'http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=", params$SRA,
  "' -O ", "data/SraMetadata_", params$SRA,
  sep = ""))

SRA_meta <- read_csv(paste("./data/SraMetadata_", params$SRA, sep = ""))

datatable(SRA_meta)
```

### List of downloaded SRA

```{r warning=FALSE, message=FALSE}
# Create SRA folder if it does not exist
if(!file_exists(paste0("mkdir -p data/", params$SRA))){
  system(paste0("mkdir -p data/", params$SRA))
}

# Check the presence of SRA already downloaded
f <- tibble(system = str_remove(basename(dir_ls(paste0("data/", params$SRA), regex = "fastq$")), 
                ".fastq.gz"))

# Create factor vectors for missing Single and Pair end SRA
missingSE <- SRA_meta$Run[(!(SRA_meta$Run %in% f)) & SRA_meta$LibraryLayout == "SINGLE" ]
missingPE <- SRA_meta$Run[(!(SRA_meta$Run %in% f)) & SRA_meta$LibraryLayout == "PAIR" ]
  

# Download the missingSE SRA
if(length(missingSE) != 0){
  for (i in seq_along(missingSE)){
    system(paste0(
      params$sratoolkit, " -e 30 -p ", missingSE[i],
      " -O data/", params$SRA))
  }
}

# Download the missingPE SRA
if(length(missingPE) != 0){
  for (i in seq_along(missingPE)){
    system(paste0(
      params$sratoolkit, " -e 30 --split-files -p ", missingPE[i],
      " -O data/", params$SRA))
  }
}


datatable(tibble(system = dir_ls(paste0("./data/", params$SRA), regex = "fastq$")))
```

### reads alignment

```{r warning=FALSE, message=FALSE}
# Read file path
f <- dir_ls(paste0("data/", params$SRA), regex = "fastq$")

########################################
# Create sketch on the original reads
for (i in f){
    system(paste0(params$bbtools,"sendsketch.sh in=", i,
                  " out=", str_replace(i, "fastq", "sketchOriginal"), ".tsv reads=1000000 ",
                  "  minprob=0.2 samplerate=1.0 merge printname0=f ", 
                  " records=20 overwrite=true color=false depth depth2",
                  " unique2 volume sortbyvolume contam2=genus nt ow" ))
}

# Parse output
so <- map_df(dir_ls(paste0("data/", params$SRA), regex = "sketchOriginal"),
          read_delim, delim = "\t", .id = "SRA", skip = 2)

datatable(so)
```

```{r warning=FALSE, message=FALSE}
########################################
# Remove Cassava genome
for (ii in f){
    system(paste0(params$bbtools, "bbmap.sh ref=_ncbi_downloads/genomes/doc_Manihot_esculenta_db_refseq_summary_statistics.tsv", 
              " in=", ii, 
              " outu=", ii, "_unmapped.fastq"))
}

########################################
# Create sketch
f <- dir_ls(paste0("data/", params$SRA), regex = "unmapped")

for (i in f){
    system(paste0(params$bbtools,"sendsketch.sh in=", i,
                  " out=", str_replace(i, "fastq$", "sketchUnmap"), ".tsv reads=1000000 ",
                  "  minprob=0.2 samplerate=1.0 merge printname0=f ", 
                  " records=20 overwrite=true color=false depth depth2",
                  " unique2 volume sortbyvolume contam2=genus nt ow" ))
}

su <- map_df(dir_ls(paste0("data/", params$SRA), regex = "sketchUnmap"),
          read_delim, delim = "\t", .id = "SRA", skip = 2)

datatable(su)
```

###TRINITY

```{r warning=FALSE, message=FALSE}
# Trinity out_file input file creation
out_file <- paste(getwd(), "/data/",params$SRA,"_trinity_samples", sep = "") 

# Path to files
fileNames <- dir_ls(paste0("data/", params$SRA), regex = "unmapped")
# write single end runs to out_file
df <- tibble(X1 = str_remove(basename(fileNames), '.fastq'),
       X2 = str_replace(basename(fileNames), '.fastq', '_rep1'),
       X3 = fileNames)


write.table(df, out_file, append = FALSE, sep = "\t", dec = ".",
            row.names = FALSE, col.names = FALSE, quote = FALSE)
```



```{r warning=FALSE, message=FALSE}

# trinity command execution 

system(paste0(params$trinity, " --seqType fq --samples_file ", out_file,
              " --CPU 40 --trimmomatic --no_normalize_reads --full_cleanup ",
              " --min_contig_length 1000 --max_memory 200G --output ",
              getwd(),"/data/",params$SRA,"/trinity_out_dir" ))

```
